cli/src/deno.rs

```diff
@@ -199,35 +206,86 @@ impl ModuleLoader for ExtensionsModuleLoader {
             };
 
             if should_transpile {
-                code = transpile(module_specifier.to_string(), code, media_type)?;
+                let transpiled =
+                    transpiler.transpile(module_specifier.to_string(), code, media_type)?;
+                code = transpiled.text.clone();
             }
 
             Ok(ModuleSource::new(module_type, code.into(), &module_specifier))
         })
     }
 }
 
-/// Transpile code to JavaScript.
-fn transpile(
-    specifier: impl Into<String>,
-    code: impl Into<String>,
-    media_type: MediaType,
-) -> Result<String> {
-    let parsed = deno_ast::parse_module(ParseParams {
-        text_info: SourceTextInfo::from_string(code.into()),
-        specifier: specifier.into(),
-        capture_tokens: false,
-        scope_analysis: false,
-        maybe_syntax: None,
-        media_type,
-    })?;
-    Ok(parsed.transpile(&Default::default())?.text)
+impl SourceMapGetter for ExtensionsModuleLoader {
+    fn get_source_map(&self, file_name: &str) -> Option<Vec<u8>> {
+        let transpiled = self.transpiler.cache.get(file_name)?;
+        transpiled.source_map.clone().map(|map| map.into_bytes())
+    }
+
+    fn get_source_line(&self, file_name: &str, line_number: usize) -> Option<String> {
+        let source_map = self.get_source_map(file_name)?;
+
+        let text = String::from_utf8(source_map).ok()?;
+
+        text.lines().nth(line_number).map(|line| line.to_owned())
+    }
+}
+
+/// Module transpilation cache.
+#[derive(Default)]
+struct Transpiler {
+    cache: DashMap<String, TranspiledSource>,
+}
+
+impl Transpiler {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    /// Transpile code to JavaScript.
+    fn transpile(
+        &self,
+        specifier: impl Into<String>,
+        code: impl Into<String>,
+        media_type: MediaType,
+    ) -> Result<TranspiledSource> {
+        let specifier = specifier.into();
+
+        // Load module if it is not in the cache.
+        if !self.cache.contains_key(&specifier) {
+            // Parse module.
+            let parsed = deno_ast::parse_module(ParseParams {
+                text_info: SourceTextInfo::from_string(code.into()),
+                specifier: specifier.clone(),
+                capture_tokens: false,
+                scope_analysis: false,
+                maybe_syntax: None,
+                media_type,
+            })?;
+
+            // Transpile to JavaScript.
+            let options = EmitOptions { inline_source_map: false, ..EmitOptions::default() };
```

Comment:
Disabling the inlining of the source map here is probably the change I'm most uncertain about, which is why I asked about ideal solutions on the Deno repo:

https://github.com/denoland/deno/issues/19617#issuecomment-1678031648

I do believe either way should work just fine though and I doubt it makes any difference for our usecase.

Reply:
For what it's worth, it looks like the deno CLI keeps the inline source map and then [extracts it from the code](https://github.com/denoland/deno/blob/625bd3905047173390eeaffd0ffcbb0bf340bb52/cli/util/text_encoding.rs#L57-L71) when needed.

Reply:
Yeah that's what I expected, which I think is definitely worse. Though it might be the only option in some scenarios?